packages/angular_devkit/build_angular/src/builders/karma/application_builder.ts

```diff
@@ -33,6 +34,76 @@ class ApplicationBuildError extends Error {
   }
 }
 
+function injectKarmaReporter(
+  context: BuilderContext,
+  buildOptions: ApplicationBuilderInternalOptions,
+  karmaConfig: Config & ConfigOptions,
+  subscriber: Subscriber<BuilderOutput>,
+) {
+  const reporterName = 'angular-progress-notifier';
+
+  interface RunCompleteInfo {
+    exitCode: number;
+  }
+
+  interface KarmaEmitter {
+    refreshFiles(): void;
+  }
+
+  class ProgressNotifierReporter {
+    static $inject = ['emitter'];
+
+    constructor(private readonly emitter: KarmaEmitter) {
+      this.startWatchingBuild();
+    }
+
+    private startWatchingBuild() {
+      void (async () => {
+        for await (const buildOutput of buildApplicationInternal(
+          {
+            ...buildOptions,
+            watch: true,
+          },
+          context,
+        )) {
+          if (buildOutput.kind === ResultKind.Failure) {
+            subscriber.next({ success: false, message: 'Build failed' });
+          } else if (
+            buildOutput.kind === ResultKind.Incremental ||
+            buildOutput.kind === ResultKind.Full
+          ) {
+            this.emitter.refreshFiles();
```

Comment:
The result files will need to be written here first so that the new test run can use them.
The integration test happens to pass now due to the syntax error that is introduced and the build failing before the tests run.

Reply:
Updated to write test files before calling `refreshFiles` - thanks for catching that!